### PR TITLE
Refactoring modal steps and creating the new Overlay modal

### PR DESCRIPTION
**Pull request by OmarG247**
- Refactored steps to be shared between modals

1. ConfirmOperationComponent
2. FileUploadComponent (previously DefaultFileUploadComponent)
3. SpecNameInput Component

- Updated directory structure to reflect shared components
- Refactored DisplayResultsComponent to display different labels according to the operation completed
- Implemented new parent Modal class which increased re-usability of code between modals
- Added the new Overlay modal for the overlay operation
- Some updates to styling

Note: I'll be updating the architecture and component diagrams to reflect this last change

### CODE REVIEW

**Review by AlanPeters — CHANGES_REQUESTED** [review]
I left a few minor comments to address.

I think that in general when looking to re-use code like we did in the modals, we should look to Composition rather than Inheritance. Inheritance can lead to some interesting bugs and overly close coupling between the components. Here is an article that touches on it: https://blog.bitsrc.io/component-reusability-techniques-with-angular-727a6c603ad2. This isn't something that we need to change now but is a design consideration worth looking at in the future. 

You have really picked up Angular during this time!

File: app/src/app/app.component.ts [hunk 1]
```diff
@@ -45,9 +50,11 @@ export class AppComponent implements AfterViewInit {
     this.mobileQuery = media.matchMedia(MOBILE_MEDIA_QUERY);
   }
 
-  openDialog() {
+  openModal(operation: string) {
```
[the remark this answers is not in the document]
Comment (OmarG247) [reply app/src/app/app.component.ts:53]: Done 👍

File: app/src/app/components/modals/modal.ts [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    toolTipText?: string;
+    nextStep?: number;
+    previousStep?: number;
+    nextButtonText?: string;
+    lastStep?: boolean;
+    stepLabel: string;
+  };
+};
+
+export class SpecMathModal {
```
[the remark this answers is not in the document]
Comment (OmarG247) [reply app/src/app/components/modals/modal.ts:39]: Renamed both class file and styling file

File: app/src/shared/services/specmath.service.ts [hunk 3]
```diff
@@ -30,15 +30,27 @@ export class SpecMathService {
...
+  /**
+   * Performs a call to the operations/overlay endpoint
+   *
+   * @param requestBody - An MergeRequest object which can contain
+   * 1. An overlay file
+   * 2. Spec used to complete the operation
+   */
+  overlaySpecs(requestBody: OverlayRequest): Observable<OperationResponse> {
+    return this.http.post(`${SPEC_MATH_URL}${routes.version}${routes.overlaySpecs}`, requestBody, requestOptions)
+      .pipe(map(response => response as any as OperationResponse));
```
[the remark this answers is not in the document]
Comment (OmarG247) [reply app/src/shared/services/specmath.service.ts]: Updated 👍

File: app/src/app/components/modals/shared-steps/file-upload/file-upload.component.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+import { MatDialog } from '@angular/material/dialog';
+
+type FileType = 'defaults' | 'spec';
+
+@Component({
+  selector: 'app-file-upload',
+  templateUrl: './file-upload.component.html',
+  styleUrls: ['./file-upload.component.scss']
+})
+export class FileUploadComponent  {
```
[the remark this answers is not in the document]
Comment (OmarG247) [reply app/src/app/components/modals/shared-steps/file-upload/file-upload.component.ts:27]: Will be creating a PR to add comments to these members

**Review by ApigeeCorky — APPROVED** [review]

**Review by AlanPeters — APPROVED** [review]

**Pull request merged**